### ISSUE TITLE
[Feature Request] Option to use HTTP/SOCKS proxy

### ISSUE THREAD

**Issue by roniemartinez** [spider-rs/spider#110]
Thanks for this awesome tool!

I would like to know if it is possible to use a HTTP/SOCKS proxy.

Cheers!

Edit: Fix

**Comment by j-mendez** [comment]
Hi 👋, @roniemartinez wow, somehow this was overlooked. I’ll add this in the morning since it’s a feature that’s a must. We have proxy support if your using a worker technically with the ‘—decentralized’ feature flag.

**Comment by j-mendez** [comment]
@roniemartinez good morning from Florida, changes are now in https://github.com/spider-rs/spider/releases/tag/v1.28.0.
Below is a basic example:

```rust
#[tokio::main]
async fn test_crawl_proxy() {
    let mut website: Website = Website::new("https://choosealicense.com");
    website
        .configuration
        .proxies
        .get_or_insert(Default::default())
        .push("socks5://184.178.172.5:15303".into());

    website.crawl().await;
}
```

In order to add `socks5` support the feature flag `socks` needs to be enabled as well.


Lmk if you have any questions and thanks again!

**Comment by roniemartinez** [comment]
@j-mendez Awesome! I will take a look and test in a few hours.

**Comment by j-mendez** [comment]
> Hi @j-mendez I am not sure what I am missing but I am not seeing the traffic passing through my socks5 proxy server.
> 
> I am verifying it using curl which shows me the logs properly.

I tested it with a http,https proxy on the net and for socks5 I used  merino “cargo install merino” and “merino —noauth” to start the proxy. The socks5 proxy uses http for the default intercept type. Could you try seeing if merino works? We may need to adjust the proxy flexibility.

**Comment by j-mendez** [comment]
Here’s the commit for the proxy change, https://github.com/spider-rs/spider/commit/1009aa9c408fb8d82fd470b437ff179198c12fb2#. I don’t mind testing the proxy if it’s OSS or available somewhere. Changing the default intercept from “http” to “https” could see if the proxy was setup secure. If it’s connecting to the secure connection we just need to add some flexibility on determining the default proxy type.

**Comment by roniemartinez** [comment]
It is also not passing through `merino`. This log only shows 1 connection which is from my `curl` request.

```
⇒  merino --no-auth         

                      _
  _ __ ___   ___ _ __(_)_ __   ___
 | '_ ` _ \ / _ \ '__| | '_ \ / _ \
 | | | | | |  __/ |  | | | | | (_) |
 |_| |_| |_|\___|_|  |_|_| |_|\___/

 A SOCKS5 Proxy server written in Rust

 2023-05-02T20:42:24.569 WARN  merino > No Authentication methods enabled. Clients will not be able to connect!
 2023-05-02T20:42:24.569 INFO  merino > Listening on 127.0.0.1:1080
 2023-05-02T20:42:24.569 INFO  merino > Serving Connections...
 2023-05-02T20:42:29.395 INFO  merino > New Request: Source: 127.0.0.1, Command: Connect Addr: 185.199.111.153, Port: 443
```

**Comment by j-mendez** [comment]
@roniemartinez is the “socks” feature flag on? It’s working on my Mac m1 and the GitHub action it looks like. I may have overlooked something, can check later today.

**Comment by roniemartinez** [comment]
To add, I am using this code for testing which shows me the visited links.

```rust
    let mut website = Website::new("https://choosealicense.com/");
    website.on_link_find_callback = Some(|s| {
        println!("link target: {}", s);
        s
    });
    website
        .configuration
        .proxies
        .get_or_insert(Default::default())
        .push("socks5://127.0.0.1:1080".to_string());

    website.crawl().await;
```

**Comment by roniemartinez** [comment]
@j-mendez 

> is the “socks” feature flag on?

Yes. Installed with `cargo add spider --features socks`

**Comment by j-mendez** [comment]
> @j-mendez
> 
> > is the “socks” feature flag on?
> 
> Yes. Installed with `cargo add spider --features socks`

What OS are you using?

**Comment by roniemartinez** [comment]
@j-mendez MacOS Ventura

```
$ sw_vers
ProductName:		macOS
ProductVersion:		13.3.1
```

**Comment by j-mendez** [comment]
> @j-mendez MacOS Ventura
> 
> ```
> $ sw_vers
> ProductName:		macOS
> ProductVersion:		13.3.1
> ```

Thanks, I think see the issue. Will post once fixed on it now.

**Comment by j-mendez** [comment]
Fixed in `v1.28.1`

```sh
 A SOCKS5 Proxy server written in Rust

 2023-05-02T15:00:13.481 WARN  merino > No Authentication methods enabled. Clients will not be able to connect!
 2023-05-02T15:00:13.481 INFO  merino > Listening on 127.0.0.1:1080
 2023-05-02T15:00:13.481 INFO  merino > Serving Connections...
 2023-05-02T15:00:30.940 INFO  merino > New Request: Source: 127.0.0.1, Command: Connect Addr: 2606:50c0:8000:0:0:0:0:153, Port: 443
```

**Comment by roniemartinez** [comment]
@j-mendez Thank you. That worked! 🚀 

Just 1 minor comment/improvement, I think you can also include the `HTTP` proxy (`http://`) to `Proxy::all()` because it can technically proxy HTTPS traffic.

I verified this by setting up only HTTP proxy in curl, Chrome and Firefox.

For reference: https://chromium.googlesource.com/chromium/src/+/HEAD/net/docs/proxy.md#http-proxy-scheme

**Comment by j-mendez** [comment]
@roniemartinez np, you are correct adding this change now. Thanks again🥳

**Comment by j-mendez** [comment]
Released in `v1.28.2` https://github.com/spider-rs/spider/commit/44eb18467d6a94b6be9ab7ff88318b5cb91d09a3

**Comment by roniemartinez** [comment]
Awesome! Thank you @j-mendez 🙇
